datafusion/common/src/tree_node.rs

```diff
@@ -366,12 +350,47 @@ impl<T: DynTreeNode + ?Sized> TreeNode for Arc<T> {
     {
         let children = self.arc_children();
         if !children.is_empty() {
-            let new_children: Result<Vec<_>> =
-                children.into_iter().map(transform).collect();
+            let new_children =
+                children.into_iter().map(transform).collect::<Result<_>>()?;
             let arc_self = Arc::clone(&self);
-            self.with_new_arc_children(arc_self, new_children?)
+            self.with_new_arc_children(arc_self, new_children)
         } else {
             Ok(self)
         }
     }
 }
+
+pub trait ConcreteTreeNode: Sized {
```

Comment:
I think we should document this trait, focusing what it should be used for, perhaps using some of the great description from this PR?

I think documenting what the functions do and their expectations, `take_children` in particular, would be valuable as well